add torch post (#103)
rstudio/ai-blog: _posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

---

> title: "Please allow me to introduce myself: torch for R"

Author: ```suggestion
title: "Please allow me to introduce myself: Torch for R"
```

---
title: "Please allow me to introduce myself: Torch for R"
description: > 

>  Today, we are excited to introduce torch, an R package allowing to use PyTorch-like functionality natively from R. No Python installation is required: torch is built directly on top of libtorch, a C++ library that provides the tensor computation and automatic differentiation capabilities essential to building neural networks.

Author: oooooh yeah
i thought i had learned that by now

Contrary to what you might expect -- if until now, you've been using `keras` -- the additional dimension is *not* the batch dimension. Batching will be taken care of by the `dataloader`, to be introduced next. Instead, this is the *channels* dimension that in `torch`, is found *before* the width and height dimensions by default.

One thing I've found to be extremely useful about `torch` is how easy it is to inspect objects. Even though we're dealing with a `dataset`, a custom object, and not an R array or even a `torch` tensor, we can easily peek at what's inside. Indexing in `torch` is 1-based, conforming to the R user's intuitions. Consequently,

```{r}
train_ds[1]
```

gives us the first element in the dataset, an R *list* of two tensors corresponding to input and target, respectively. (We don't reproduce the output here, but you can see for yourself in the notebook.)

Let's inspect the shape of the input tensor:

```{r}
train_ds[1][[1]]$size()
```

    [1]  1 28 28

Now we have the data, we need someone to feed them to a deep learning model, nicely batched and all. In `torch`, this is the task of data loaders.

#### Data loader

> Each of training and test sets get their own data loader:

Author: yeah i think i didn't know

```{r}
net <- nn_module(
  
  "KMNIST-CNN",
  
  initialize = function() {
    # in_channels, out_channels, kernel_size, stride = 1, padding = 0
    self$conv1 <- nn_conv2d(1, 32, 3)
    self$conv2 <- nn_conv2d(32, 64, 3)
    self$dropout1 <- nn_dropout2d(0.25)
    self$dropout2 <- nn_dropout2d(0.5)
    self$fc1 <- nn_linear(9216, 128)
    self$fc2 <- nn_linear(128, 10)
  },
  
  forward = function(x) {
    x %>% 
      self$conv1() %>%
      nnf_relu() %>%
      self$conv2() %>%
      nnf_relu() %>%
      nnf_max_pool2d(2) %>%
      self$dropout1() %>%
      torch_flatten(start_dim = 2) %>%
      self$fc1() %>%
      nnf_relu() %>%
      self$dropout2() %>%
      self$fc2()
  }
)
```

The layers -- apologies: modules -- themselves may look familiar. Unsurprisingly, `nn_conv2d()` performs two-dimensional convolution; `nn_linear()` multiplies by a weight matrix and adds a vector of biases. But what are those numbers: `nn_linear(128, 10)`, say?

> In `torch`, instead of the number of units in a layer, you specify input and output dimensionalities of the "data" that run through it. Thus, `nn_linear(128, 10)` has 128 input connections and outputs 10 values -- one for every class. In some cases, such as this one, specifying dimensions is easy -- we know how many input edges there are (namely, the same as the number of output edges from the previous layer), and we know how many output values we need. But how about this module's predecessor? How do we arrive at `9216` input connections?

Reviewer: I'm a little confused about this:
> But how about this module's predecessor? How do we arrive at `9216` input connections?

OK, I just un-confused myself by scrolling back up in the code, and I see that you're referring to:
```r
    self$fc1 <- nn_linear(9216, 128)
    self$fc2 <- nn_linear(128, 10)
```

I think it might be easier to follow if you reference them as `fc1` and `fc2` along with giving the dimensions. Alone I wasn't totally sure what you meant by "predecessor"—I guess I was thinking of it as the preceding step.

Author: i think i should change to "preceding module"

> Here, a bit of calculation is necessary. We go through all actions that happen in `forward()` -- if they change shapes, we keep track of the transformation; if they don't, we ignore them.

Reviewer: Is the "they" in "if they change shapes" meant to refer to the actions? Right now it does, but I'm not sure if you mean that the actions change shapes, or that the actions change the shape of the tensors.

Author: changing to _affect_

So, we start with input tensors of shape `batch_size x 1 x 28 x 28`. Then,

> -   `nn_conv2d(1, 32, 3)` , or equivalently, `nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3),`applies a convolution with kernel size 3, stride 1 (the default) and no padding (the default). We can consult the [documentation](https://mlverse.github.io/torch/reference/nn_conv2d.html) to look up the resulting output size, or just intuitively reason that with a kernel of size 3 and no padding, the image will shrink by one pixel in each direction, resulting in a spatial resolution of `26 x 26`. *Per channel*, that is. Thus, the actual output shape is `batch_size x 32 x 26 x 26` . Next,

Reviewer: ```suggestion
-   `nn_conv2d(1, 32, 3)` , or equivalently, `nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3),`applies a convolution with kernel size 3, stride 1 (the default), and no padding (the default). We can consult the [documentation](https://mlverse.github.io/torch/reference/nn_conv2d.html) to look up the resulting output size, or just intuitively reason that with a kernel of size 3 and no padding, the image will shrink by one pixel in each direction, resulting in a spatial resolution of `26 x 26`. *Per channel*, that is. Thus, the actual output shape is `batch_size x 32 x 26 x 26` . Next,
```

Reviewer: Added a serial comma above (doesn't show up nicely)

Finally, let's evaluate model performance on the test set.

## Evaluation

Putting a model in `eval` mode tells `torch` *not* to calculate gradients and perform backprop during the operations that follow:

```{r}
model$eval()
```

We iterate over the test set, keeping track of losses and accuracies obtained on the batches.

```{r}
test_losses <- c()
total <- 0
correct <- 0

for (b in enumerate(test_dl)) {
  output <- model(b[[1]]$to(device = "cuda"))
  labels <- b[[2]]$to(device = "cuda")
  loss <- nnf_cross_entropy(output, labels)
  test_losses <- c(test_losses, loss$item())
  # torch_max returns a list, with position 1 containing the values 
  # and position 2 containing the respective indices
  predicted <- torch_max(output$data(), dim = 2)[[2]]
  total <- total + labels$size(1)
  # add number of correct classifications in this batch to the aggregate
  correct <- correct + (predicted == labels)$sum()$item()
}

mean(test_losses)
```

    [1] 1.53784480643349

Here is mean accuracy, computed as proportion of correct classifications:

```{r}
test_accuracy <-  correct/total
test_accuracy
```

    [1] 0.9449

That's it for our first `torch` example. Where to from here?

## Learn

> To learn more, check out our vignettes on the `torch` [website](https://mlverse.github.io/torch). To begin, you may want to check out these in particular:

Reviewer: ```suggestion
To learn more, check out our vignettes on the [`torch` website](https://mlverse.github.io/torch). To begin, you may want to check out these in particular:
```

Reviewer: Just better for accessibility to have the name of the website inside of the link 😉

-   "Getting started" series: Build a simple neural network from scratch, starting from [low-level tensor manipulation](https://mlverse.github.io/tohttps://mlverse.github.io/torch/articles/getting-started/tensors.html) and gradually adding in higher-level features like [automatic differentiation](https://mlverse.github.io/torch/articles/getting-started/tensors-and-autograd.html) and [network modules](https://mlverse.github.io/torch/articles/getting-started/nn.html).
-   More on tensors: [Tensor creation](https://mlverse.github.io/torch/articles/tensor-creation.html) and [indexing](https://mlverse.github.io/torch/articles/indexing.html)
-   Backpropagation in `torch`: [autograd](https://mlverse.github.io/torch/articles/using-autograd.html)

> If you have questions, or run into problems, please feel free to ask on [GitHub](https://github.com/mlverse/torch) or on the [RStudio community forum](https://community.rstudio.com/).

Reviewer: ```suggestion
If you have questions, or run into problems, please feel free to ask on [GitHub](https://github.com/mlverse/torch) or on the [RStudio community forum](https://community.rstudio.com/c/ml/15).
```

Reviewer: ^^ links them directly to the machine learning and modelling category

Author: awesome

## We need you

> We very much wish that the R community will find the new functionality useful. But that's not all. Instead, we hope that you, many of you, will take part in the journey.

Reviewer: ```suggestion
We very much hope that the R community will find the new functionality useful. But that's not all. Instead, we hope that you, many of you, will take part in the journey.
```

Reviewer: Same comment with respect to this part as in PR 334 (instead feels off).

---
title: "Please allow me to introduce myself: torch for R"
description: > 
 Today, we are excited to introduce torch, an R package allowing to use PyTorch-like functionality natively from R. No Python installation is required: torch is built directly on top of libtorch, a C++ library that provides the tensor computation and automatic differentiation capabilities essential to building neural networks.

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanaintroducingtorchforR
bibliography: bibliography.bib
date: 09-29-2020
categories:
  - Packages/Releases
  - Torch
  - R
output:
  distill::distill_article:
    self_contained: false
preview: images/torch.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

> Last January at *rstudio::conf*, in that distant past when conferences still used to take place at some physical location, my colleague Daniel gave a talk introducing new features and ongoing development in the `tensorflow` ecosystem. In the Q&A part, he was asked something unexpected: Were we going to build support for PyTorch? He hesitated; that was in fact the plan, and he had already played around with natively implementing `torch` tensors at a prior time, but he was not completely certain how well "it" would work.

Reviewer 2: ```suggestion
Last January at [rstudio::conf](https://rstudio.com/conference/), in that distant past when conferences still used to take place at some physical location, my colleague [Daniel](https://twitter.com/dfalbel) gave a talk introducing new features and ongoing development in the `tensorflow` ecosystem. In the Q&A part, he was asked something unexpected: Were we going to build support for [PyTorch](https://pytorch.org/)? He hesitated; that was in fact the plan, and he had already played around with natively implementing `torch` tensors at a prior time, but he was not completely certain how well "it" would work.
```

"It", that is an implementation which does not bind to Python Torch, meaning, we don't install the PyTorch wheel and import it via `reticulate`. Instead, we delegate to the underlying C++ library `libtorch` for tensor computations and automatic differentiation, while neural network features -- layers, activations, optimizers -- are implemented directly in R. Removing the intermediary has at least two benefits: For one, the leaner software stack means fewer possible problems in installation and fewer places to look when troubleshooting. Secondly, through its non-dependence on Python, `torch` does not require users to install and maintain a suitable Python environment. Depending on operating system and context, this can make an enormous difference: For example, in many organizations employees are not allowed to manipulate privileged software installations on their laptops.

> So why did Daniel hesitate, and, if I recall correctly, give a not-too-conclusive answer? On the one hand, it was not clear whether compilation against `libtorch` would, on some operating systems, pose severe difficulties. (It did, but difficulties turned out to be surmountable; at some point, my colleague and `torch`-coauthor Javier might want to tell you about it.)[^1] On the other, the sheer amount of work involved in re-implementing -- not all, but a big amount of -- PyTorch in R seemed intimidating. Today, there is still lots of work to be done (we'll pick up that thread at the end), but the main obstacles have been ovecome, and enough components are available that `torch` can be useful to the R community. Thus, without further ado, let's train a neural network.
> 
> [^1]: Lest there be any misunderstandings -- coauthor not to yours truly, but to Daniel, creator and main author of the package.

Reviewer 2: Not sure if I'll ever write a post in the RStudio blog about lantern, but I think we can give a couple pointers to those interested!

```suggestion
So why did Daniel hesitate, and, if I recall correctly, give a not-too-conclusive answer? On the one hand, it was not clear whether compilation against `libtorch` would, on some operating systems, pose severe difficulties. (It did, but difficulties turned out to be surmountable; [Javier](https://twitter.com/javierluraschi) wrapped `libtorch` into [lantern](https://github.com/mlverse/lantern), a C interface to `libtorch` which avoids cross-compiler issues between MinGW and Visual Studio in Windows) On the other, the sheer amount of work involved in re-implementing -- not all, but a big amount of -- PyTorch in R seemed intimidating. Today, there is still lots of work to be done (we'll pick up that thread at the end), but the main obstacles have been ovecome, and enough components are available that `torch` can be useful to the R community. Thus, without further ado, let's train a neural network.
```

Author: hm... ok.... I think the alternative is too detailed though ... I'll see if I can "compress" it and/or move it to a footnote...

You're not at your laptop now? Just follow along in the [companion notebook on Colaboratory](https://colab.research.google.com/drive/1NdiN9n_a7NEvFpvjPDvxKTshrSWgxZK5?usp=sharing).

## Installation

#### `torch`

Installing `torch` is as straightforward as typing

```{r}
install.packages("torch")
```

This will detect whether you have CUDA installed, and either download the CPU or the GPU version of `libtorch`. Then, it will install the R package from CRAN. To make use of the very newest features, you can also install the development version from GitHub:

```{r}
devtools::install_github("mlverse/torch")
```

> To quickly check the installation, and whether GPU support works fine (assuming that there *is* a GPU), create a tensor *on the CUDA device*:

Reviewer 3: We might want to say that a GPU means a CUDA capable NVIDIA GPU

Author: right, there's more than Nvidia ...

```{r}
torch_tensor(1, device = "cuda")
```

    torch_tensor 
     1
    [ CUDAFloatType{1} ]

If all our *hello torch* example did was running a network on, say, simulated data, we could stop here. As we'll do image classification, however, we need to install another package: `torchvision`.

#### `torchvision`

```{r}
```

Whereas `torch` is where tensors, network modules, and generic data loading functionality live, datatype-specific capabilities are -- or will be -- provided by dedicated packages. In general, these capabilities comprise three types of things: datasets, tools for pre-processing and data loading, and pre-trained models.

As of this writing, PyTorch has dedicated libraries for three domain areas: vision, text, and audio. In R, we plan to proceed analogously -- "plan", because `torchtext` and `torchaudio` are still to be created. Right now, `torchvision` is all we need:

```{r}
install.packages("torchvision")
```

> Note: As `torchvision` relies on [magick](https://cran.r-project.org/web/packages/magick/vignettes/intro.html), which itself needs [ImageMagick](https://imagemagick.org/) installed, you need to ensure the latter is available on your system. [The magick vignette](https://cran.r-project.org/web/packages/magick/vignettes/intro.html) has an excellent description on how to do this on different operating systems.

Author: the way I understood it, ImageMagick per se is required on every system, no? Just that on linux, as the R package `magick` is compiled from source, you need the headers?

And we're ready to load the data.

## Data loading and pre-processing

```{r}
library(torch)

> library(torchvision)

Author: I think `torch` imports it ... in any case it works without...

```

The list of vision datasets bundled with PyTorch is long, and they're continually being added to `torchvision`.

The one we need right now is available already, and it's -- MNIST? ... not quite: It's my favorite "MNIST dropin", [Kuzushiji-MNIST](https://github.com/rois-codh/kmnist) [@clanuwat2018deep]. Like other datasets explicitly created to replace MNIST, it has ten classes -- characters, in this case, depicted as grayscale images of resolution `28x28`.

Here are the first 32 characters:

```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Kuzushiji MNIST."}
knitr::include_graphics("images/kmnist.png")
```

#### Dataset

The following code will download the data, separately for training and test sets.

```{r}
dir <- "."

train_ds <- kmnist_dataset(
  dir,
  download = TRUE,
  train = TRUE,
  transform = function(x) {

>     x <- x$to(dtype = torch_float())/256

Reviewer 3: there's a `transform_to_tensor` that takes a magick image, normalizes it into [0,1] and adds the channels dimension.

Author: right, we have that now... will change

Let's inspect the shape of the input tensor:

```{r}
train_ds[1][[1]]$size()
```

    [1]  1 28 28

Now we have the data, we need someone to feed them to a deep learning model, nicely batched and all. In `torch`, this is the task of data loaders.

#### Data loader

Each of training and test sets get their own data loader:

```{r}
train_dl <- dataloader(train_ds, batch_size = 32, shuffle = TRUE)
test_dl <- dataloader(test_ds, batch_size = 32)
```

Again, `torch` makes it easy to verify we did the correct thing. To take a look at the content of the first batch, do

```{r}
train_iter <- train_dl$.iter()
train_iter$.next()
```

Functionality like this may not seem indispensable when working with a well-known dataset, but will turn out to be very useful when a lot of domain-specific pre-processing is required.

Now that we've seen how to load data, all prerequisites are fulfilled for visualizing them. Here is the code that was used to display the first batch of characters, above:

```{r, fig.asp=0.5, fig.width=8}
par(mfrow = c(4,8), mar = rep(0, 4))
images <- train_dl$.iter()$.next()[[1]][1:32, 1, , ] 
images %>%
  purrr::array_tree(1) %>%
  purrr::map(as.raster) %>%

>   purrr::iwalk(~{plot(.x)})

Reviewer 3: Looks like the image is missing in the preview

Author: it shouldn't be there - that's why I say
> Here is the code that was used to display the first batch of characters, above:

-   `nn_linear(9216, 128)` discussed above. Again,

-   `nnf_relu()` and `nn_dropout2d(0.5)` leave dimensions as they are, and finally,

-   `nn_linear(128, 10)` gives us the desired output scores, one for each of the ten classes.

Now you'll be thinking, -- what if my network is more complicated? Calculations could become pretty cumbersome. Luckily, with `torch`'s flexibility, there is another way. Since every layer is callable *in isolation*, we can just ... create some sample data and see what happens!

Here is a sample "image" -- or more precisely, a one-item batch containing it:

```{r}
x <- torch_randn(c(1, 1, 28, 28))
```

What if we call the first *conv2d* module on it?

```{r}
conv1 <- nn_conv2d(1, 32, 3)
conv1(x)$size()
```

    [1]  1 32 26 26

Or both *conv2d* modules?

```{r}
conv2 <- nn_conv2d(32, 64, 3)
(conv1(x) %>% conv2())$size()
```

    [1]  1 64 24 24

And so on. This is just one example illustrating how `torch`s flexibility makes developing neural nets easier.

Back to the main thread. We instantiate the model, and we ask `torch` to allocate its weights (parameters) on the GPU:

```{r}

> model <- net()$to(device = "cuda")

Reviewer 3: I would split this into 2 lines. One to initialize the model and the other to move to the GPU.

Author: makes sense
